Profile Connection Manager API Enhancement

zhongzhijie1 opened the pull request:
## Profile Connection Manager API Enhancement

### Summary
This PR refactors the profile connection manager to support multi-connection instances per profile and adds user_data field for better connection context management.

### Key Changes

#### 1. Core API Enhancement (`6fb6d78`, `a4df9ed`)
- **Added `conn_id` parameter**: Enables multiple connection instances per profile
- **Added `user_data` parameter**: Allows passing connection-specific context through callbacks
- **Removed flag-based state tracking**: Replaced with `is_busy` state per connection node
- **Updated handler signature**: `bt_profile_conn_handler_t` now accepts `void* user_data`

```c
// Before
bt_status_t (*bt_profile_conn_handler_t)(bt_controller_id_t id, bt_address_t* addr);

// After
bt_status_t (*bt_profile_conn_handler_t)(bt_controller_id_t id, bt_address_t* addr, void* user_data);
```

#### 2. Profile Adaptations
- **A2DP** (`ebad02c`): Adapted source and sink profiles to new API
- **Adapter** (`7d6eb45`): Updated ACL connection callbacks with conn_id
- **AVRCP** (`1f50220`): Updated controller and target profiles
- **HID Device** (`abf8cca`): Added multi-connection support and user_data handling
- **SPP** (`75ef118`): Added conn_id field and integrated with connection manager

#### 3. Implementation Details

**Connection Manager (`a4df9ed`)**:
- Each handler node now represents a unique connection instance
- Connection identified by `(profile_id, conn_id)` tuple
- Handler lookup optimized with composite key matching
- Automatic ACL disconnect when all profiles are closed

**Profile Integration Pattern**:
```c
// Connection
bt_sal_cm_profile_connected_callback(cm_data_new(&addr, PROFILE_XXX, conn_id));
bt_sal_profile_disconnect_register(&addr, PROFILE_XXX, conn_id, PRIMARY_ADAPTER, 
                                   disconnect_handler, conn_context);

// Disconnection  
bt_sal_cm_profile_disconnected_callback(cm_data_new(&addr, PROFILE_XXX, conn_id));
```

### Benefits
1. **Multiple Connections**: Supports multiple connections per profile (e.g., multiple SPP channels)
2. **Better Context Management**: User data eliminates need for global state lookup
3. **Cleaner Code**: Removed complex flag-based state machine
4. **Type Safety**: Proper connection instance identification

### Testing
- All adapted profiles maintain backward compatibility
- Connection/disconnection flows tested with multi-instance scenarios

---

**Signed-off-by:** zhongzhijie1 <zhongzhijie1@xiaomi.com>

zhongzhijie1 committed:
sal_connection_manager API support user_data field and multi connection
instance para.

bug: v/78961

add user_data and conn_id per profile feature.

Signed-off-by: zhongzhijie1 <zhongzhijie1@xiaomi.com>

zhongzhijie1 committed:
implement user_data and multi connection instannce per profile manager

bug: v/78961

remove redundant flag based judge logic, instead of conn node's busy
state.

Each handler node represents a conn instance.

User_data can be transmitted through the cb Profile register, and the conn_id and profile_id are used to identify the conn instance of a specific profile.

Signed-off-by: zhongzhijie1 <zhongzhijie1@xiaomi.com>

zhongzhijie1 committed:
adapt a2dp profile to latest multi conn instance and user data API.

bug: v/78961

Signed-off-by: zhongzhijie1 <zhongzhijie1@xiaomi.com>

zhongzhijie1 committed:
adapt adapter interface to latest profile connection manager API.

bug: v/78961

Signed-off-by: zhongzhijie1 <zhongzhijie1@xiaomi.com>

zhongzhijie1 committed:
adapt avrcp profile to latest profile connection manager API.

bug: v/78961

Signed-off-by: zhongzhijie1 <zhongzhijie1@xiaomi.com>

chengkai15 commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -22,71 +22,59 @@
 static bt_list_t* bt_sal_disconnecting_list = NULL;
 
 static bt_status_t bt_sal_trigger_profile_conn_act(bt_profile_connection_manager_t* manager, bt_list_t* manager_list);
+static void remove_from_connection_manager_list(bt_list_t* list, bt_address_t* addr, uint8_t profile_id,
```
could we user more frendly function name， likes bt_sal_remove_connection

chengkai15 commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -260,8 +244,7 @@ void cm_data_destory(cm_data_t* data)
     bt_list_node_t* node;
-    bt_profile_conn_handler_node_t* entry_node;
-    uint64_t bit_flag;
+    bt_profile_conn_handler_node_t* handler_node;
```
conn_handler

chengkai15 commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -20,32 +20,36 @@
     bt_address_t* addr, uint8_t reason);
 
-cm_data_t* cm_data_new(bt_address_t* addr, uint8_t profile_id);
+cm_data_t* cm_data_new(bt_address_t* addr, uint8_t profile_id, uint16_t conn_id);
```
could we define default conn_id with a marco ?

huangyulong3 replied:
Why we need conn_id? `addr` include the `conn_id` already?

huangyulong3 commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -22,71 +22,59 @@
 typedef struct {
     bt_address_t device_addr;
     uint8_t profile_id;
+    uint16_t conn_id;
```
Why we need to add the `conn_id` here?

zhongzhijie1 replied:
profile such as spp, have multiple connection_instance, so we need a conn_id parameter to distinguish

huangyulong3 reviewed (commented):
LGTM.

huangyulong3 reviewed (approved):
LGTM.

huangyulong3 reviewed (approved):
LGTM.

huangyulong3 reviewed (approved):
LGTM.

huangyulong3 commented on service/stacks/zephyr/sal_spp_interface.c:
```diff
@@ -677,16 +686,61 @@ static bt_status_t spp_connect_with_uuid(sal_spp_connection_t* spp_conn, bt_uuid
+
+    spp_conn->conn = conn;
+
+    if (spp_conn->conn_port & 0x3F) {
```
What is the `0x3F` meaning? Can we don't used the magic number?

zhongzhijie1 replied:
it just copy the original spp connect function implement

huangyulong3 commented on service/stacks/zephyr/sal_spp_interface.c:
```diff
@@ -677,16 +686,61 @@ static bt_status_t spp_connect_with_uuid(sal_spp_connection_t* spp_conn, bt_uuid
+    spp_conn->conn = conn;
+
+    if (spp_conn->conn_port & 0x3F) {
+        int err;
```
Can we define the `int err` in the begining,don't define it again in the `else` case.

zhongzhijie1 replied:
it just copy the original spp connect function implement

zhongzhijie1 committed:
use micro CONN_ID_DEFAULT 0x0001 instead of number 0.

bug: v/78961

Signed-off-by: zhongzhijie1 <zhongzhijie1@xiaomi.com>

gzh-terry commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -17,35 +17,41 @@
 typedef struct {
     bt_profile_conn_handler_t handler;
     uint8_t profile_id;
+    uint16_t conn_id;
```
Should we use `user_data` to distinguish between different requests?
If required by the upper layer, the `conn_id` can be placed into `user_data`.
This module's concern should be limited to the profile; a single profile can be associated with multiple `user_data`.

zhongzhijie1 replied:
想过这样搞，但是 user data 是指针，语义上给业务 passthrough， id 天然有编号 ，加上 profile_id 可以识别唯一的连接 object

gzh-terry commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -17,35 +17,41 @@
     bt_profile_conn_handler_t handler;
     uint8_t profile_id;
+    uint16_t conn_id;
+    bool is_busy;
```
What's the life cycle and the purpose for `is_busy`?

zhongzhijie1 replied:
防止重复调用连接，和之前那个 profile_flag 作用一样

zhongzhijie1 committed:
support hfp hf profile connect interface.

bug: v/77997

Signed-off-by: zhongzhijie1 <zhongzhijie1@xiaomi.com>

chengkai15 committed:
Bluetooth: SPP: Add ACL connection manager feature

bug: v/79672

Add ACL connection manager feature, include acl
connect request feature, and many conenctions
feature.

Signed-off-by: Kai Cheng <chengkai@xiaomi.com>

chengkai15 committed:
Bluetooth: HID: Add ACL connection manager feature

bug: v/79672

Add ACL connection manager feature, include acl
connect request feature, and many conenctions
feature.

Signed-off-by: Kai Cheng <chengkai@xiaomi.com>

liuX10 reviewed (commented):
LGTM

liuX10 reviewed (commented):
add commit msg

liuX10 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
add commit msg for commit:"support hfp hf profile connect interface."

liuX10 commented on service/stacks/zephyr/include/sal_connection_manager.h:
add commit msg for commit:"use micro CONN_ID_DEFAULT 0x0001 instead of number 0."

liuX10 commented on service/stacks/zephyr/sal_avrcp_interface.c:
add commit msg for commit:"adapt avrcp profile to latest profile connection manager API."

liuX10 commented on service/stacks/zephyr/sal_adapter_interface.c:
add commit msg for commit:"adapt adapter interface to latest profile connection manager API."

liuX10 commented on service/stacks/zephyr/sal_a2dp_interface.c:
add commit msg for commit:"adapt a2dp profile to latest multi conn instance and user data API."
